### PR TITLE
Harden how ConstEval uses llvm-cpu and the runtime libraries.

### LINKED ISSUE #17070: Cmake blob incompleteness
### What happened?

The console output was the following:

```bash
root@5f05eee1f19a:/npu-mlir-compiler# iree-compile --iree-hal-target-backends=llvm-cpu mobilenetv2.mlir -o mobilenetv2_cpu.vmfb
<vm>:0: UNAVAILABLE; none of the executable binaries in the module are supported by the runtime;
[ 0] bytecode module.__init:246 -:1:1
Please report issues to https://github.com/openxla/iree/issues and include the crash backtrace.
Stack dump:
0.      Program arguments: iree-compile --iree-hal-target-backends=llvm-cpu mobilenetv2.mlir -o mobilenetv2_cpu.vmfb
 #0 0x00007ff37059c498 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:13
 #1 0x00007ff37059a6b0 llvm::sys::RunSignalHandlers() /npu-mlir-compiler/third-party/iree/third_party/llvm-project/llvm/lib/Support/Signals.cpp:106:18
 #2 0x00007ff37059cb38 SignalHandler(int) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/llvm/lib/Support/Unix/Signals.inc:413:1
 #3 0x00007ff36b8a2520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007ff36b8f69fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #5 0x00007ff36b8a2476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #6 0x00007ff36b8887f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #7 0x00007ff373237849 /npu-mlir-compiler/third-party/iree/runtime/src/iree/base/assert.h:26:63
 #8 0x00007ff37323778c /npu-mlir-compiler/third-party/iree/runtime/src/iree/base/status.c:552:3
 #9 0x00007ff370f5c714 mlir::iree_compiler::ConstEval::Runtime::getInstance() /npu-mlir-compiler/third-party/iree/compiler/src/iree/compiler/ConstEval/Runtime.cpp:494:3
#10 0x00007ff370f5c714 mlir::iree_compiler::ConstEval::CompiledBinary::initialize(void*, unsigned long) /npu-mlir-compiler/third-party/iree/compiler/src/iree/compiler/ConstEval/Runtime.cpp:431:22
#11 0x00007ff370f5c9da mlir::iree_compiler::ConstEval::InMemoryCompiledBinary::translateFromModule(mlir::ModuleOp) /npu-mlir-compiler/third-party/iree/compiler/src/iree/compiler/ConstEval/Runtime.cpp:0:3
#12 0x00007ff3706d0cfd mlir::LogicalResult::failed() const /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/include/mlir/Support/LogicalResult.h:44:33
#13 0x00007ff3706d0cfd mlir::failed(mlir::LogicalResult) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/include/mlir/Support/LogicalResult.h:72:58
#14 0x00007ff3706d0cfd mlir::iree_compiler::ConstEval::(anonymous namespace)::JitGlobalsPass::runOnOperation() /npu-mlir-compiler/third-party/iree/compiler/src/iree/compiler/ConstEval/JitGlobals.cpp:548:9
#15 0x00007ff37072e5df mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_1::operator()() const /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/lib/Pass/Pass.cpp:0:17
#16 0x00007ff37072e5df void llvm::function_ref<void ()>::callback_fn<mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_1>(long) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:45:12
#17 0x00007ff37072e5df llvm::function_ref<void ()>::operator()() const /npu-mlir-compiler/third-party/iree/third_party/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:68:12
#18 0x00007ff37072e5df void mlir::MLIRContext::executeAction<mlir::PassExecutionAction, mlir::Pass&>(llvm::function_ref<void ()>, llvm::ArrayRef<mlir::IRUnit>, mlir::Pass&) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/include/mlir/IR/MLIRContext.h:275:7
#19 0x00007ff37072e5df mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/lib/Pass/Pass.cpp:521:21
#20 0x00007ff37072ed69 mlir::LogicalResult::failed() const /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/include/mlir/Support/LogicalResult.h:44:33
#21 0x00007ff37072ed69 mlir::failed(mlir::LogicalResult) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/include/mlir/Support/LogicalResult.h:72:58
#22 0x00007ff37072ed69 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/lib/Pass/Pass.cpp:593:9
#23 0x00007ff37073351f mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (mlir::OpPassManager&, mlir::Operation*)>::callback_fn<mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_0>(long, mlir::OpPassManager&, mlir::Operation*) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:45:5
#24 0x00007ff3716dc102 mlir::iree_compiler::IREE::Util::(anonymous namespace)::FixedPointIteratorPass::runOnOperation() /npu-mlir-compiler/third-party/iree/compiler/src/iree/compiler/Dialect/Util/Transforms/FixedPointIterator.cpp:0:0
#25 0x00007ff37072e5df mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_1::operator()() const /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/lib/Pass/Pass.cpp:0:17
#26 0x00007ff37072e5df void llvm::function_ref<void ()>::callback_fn<mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int)::$_1>(long) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:45:12
#27 0x00007ff37072e5df llvm::function_ref<void ()>::operator()() const /npu-mlir-compiler/third-party/iree/third_party/llvm-project/llvm/include/llvm/ADT/STLFunctionalExtras.h:68:12
#28 0x00007ff37072e5df void mlir::MLIRContext::executeAction<mlir::PassExecutionAction, mlir::Pass&>(llvm::function_ref<void ()>, llvm::ArrayRef<mlir::IRUnit>, mlir::Pass&) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/include/mlir/IR/MLIRContext.h:275:7
#29 0x00007ff37072e5df mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/lib/Pass/Pass.cpp:521:21
#30 0x00007ff37072ed69 mlir::LogicalResult::failed() const /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/include/mlir/Support/LogicalResult.h:44:33
#31 0x00007ff37072ed69 mlir::failed(mlir::LogicalResult) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/include/mlir/Support/LogicalResult.h:72:58
#32 0x00007ff37072ed69 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/lib/Pass/Pass.cpp:593:9
#33 0x00007ff370731279 mlir::PassManager::run(mlir::Operation*) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/lib/Pass/Pass.cpp:884:0
#34 0x00007ff3704f5eae mlir::LogicalResult::failed() const /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/include/mlir/Support/LogicalResult.h:44:33
#35 0x00007ff3704f5eae mlir::failed(mlir::LogicalResult) /npu-mlir-compiler/third-party/iree/third_party/llvm-project/mlir/include/mlir/Support/LogicalResult.h:72:58
#36 0x00007ff3704f5eae mlir::iree_compiler::embed::(anonymous namespace)::Invocation::runPipeline(iree_compiler_pipeline_t) /npu-mlir-compiler/third-party/iree/compiler/src/iree/compiler/API/Internal/CompilerDriver.cpp:997:7
#37 0x00007ff3704f5eae ireeCompilerInvocationPipeline /npu-mlir-compiler/third-party/iree/compiler/src/iree/compiler/API/Internal/CompilerDriver.cpp:1432:23
#38 0x00007ff3706f85f8 mlir::iree_compiler::runIreecMain(int, char**)::$_2::operator()(iree_compiler_source_t*) const /npu-mlir-compiler/third-party/iree/compiler/src/iree/compiler/Tools/iree_compile_lib.cc:254:11
#39 0x00007ff3706f7e79 mlir::iree_compiler::runIreecMain(int, char**) /npu-mlir-compiler/third-party/iree/compiler/src/iree/compiler/Tools/iree_compile_lib.cc:0:10
#40 0x00007ff36b889d90 (/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
#41 0x00007ff36b889e40 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e40)
#42 0x00005626296a06a5 _start (/npu-mlir-compiler/third-party/iree-build/tools/iree-compile+0x16a5)
Aborted
```

As suggested by `Stella Laurenzo` on the IREE discord, adding the following command line parameter allowed the compilation to go through successfully:

```bash
--iree-opt-const-eval=false
```

### Steps to reproduce your issue

I generated a model with the following script in python:

```python
import tensorflow.compat.v2 as tf
import tensorflow_hub as hub

HUB_PATH = "https://tfhub.dev/google/tf2-preview/mobilenet_v2/classification/4"
model_path = hub.resolve(HUB_PATH)
print(f"Downloaded model from tfhub to path: '{model_path}'")

loaded_model = tf.saved_model.load(model_path)
call = loaded_model.__call__.get_concrete_function(tf.TensorSpec([1, 224, 224, 3], tf.float32))

resaved_model_path = 'MobileNetV2.tf'
tf.saved_model.save(loaded_model, resaved_model_path, signatures={'predict': call})
```

Then i invoked the following script to convert the .tf model into a .mlir file:
```bash
iree-import-tf \
--tf-import-type=savedmodel_v1 \
--tf-savedmodel-exported-names=predict \ 
MobileNetV2.tf/ -o mobilenetv2.mlir
```

And, finally, this line was invoked to get the FlatBuffers file from the MLIR representation of the model:
```bash
iree-compile --iree-hal-target-backends=llvm-cpu mobilenetv2.mlir -o mobilenetv2_cpu.vmfb
```

which resulted in the described error

### What component(s) does this issue relate to?

Compiler

### Version information

```bash
root@5f05eee1f19a:/npu-mlir-compiler# iree-compile --version
IREE (https://iree.dev):
  IREE compiler version (unknown)
  LLVM version 19.0.0git
  Optimized build with assertions
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ScottTodd**
* Fixed https://github.com/openxla/iree/issues/17070 by updating the CMake options needed for ConstEval
* Replaced `IREE_CHECK_OK` usage with error handling
* Refactored `test/jit_globals.mlir`, adding coverage for llvm-cpu (since that is actually running by default)
  * I tried to keep all test cases in one file, but `--verify-diagnostics` isn't compatible with that style of lit testing AFAICT
  * This uncovered some bugs in https://github.com/openxla/iree/pull/16321 / missing support for i4 types

### CODE REVIEW

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 #include "iree/compiler/ConstEval/Runtime.h"
 
+#include "iree/base/status_cc.h"
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/Runtime.cpp]: include iree/base/api.h - status_cc is included automatically

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp [hunk 2]
```diff
@@ -149,13 +141,19 @@ void CompiledBinary::deinitialize() {
...
-                                    argCapacity, iree_allocator_system(),
-                                    &inputs));
-  IREE_CHECK_OK(iree_vm_list_create(iree_vm_make_undefined_type_def(),
-                                    resultCapacity, iree_allocator_system(),
-                                    &outputs));
+    : binary(binary), argCapacity(argCapacity), resultCapacity(resultCapacity) {
+}
+
+LogicalResult FunctionCall::initialize() {
+  if (!iree_status_is_ok(iree_vm_list_create(iree_vm_make_undefined_type_def(),
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/Runtime.cpp]: you can't nest status creation with isok as you leak the status
you need to use a macro that does the right thing or a function that converts and consumes for you - you'd iree_status_ignore or ideally log and then free before returning your LogicalResult. You can use InFlightDiagnostic and such to append the status and return the failure.

I think handleRuntimeError does most of that

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp [hunk 3]
```diff
@@ -149,13 +141,19 @@ void CompiledBinary::deinitialize() {
...
+                                             argCapacity,
+                                             iree_allocator_system(), &inputs)))
+    return failure();
+  if (!iree_status_is_ok(
+          iree_vm_list_create(iree_vm_make_undefined_type_def(), resultCapacity,
+                              iree_allocator_system(), &outputs)))
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/Runtime.cpp]: these should probably be default nowadays? or better, properly captured so we have only a single place that decides and the rest are `iree_allocator_t host_allocator` members or arguments

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp [hunk 4]
```diff
@@ -472,17 +498,20 @@ InMemoryCompiledBinary::translateFromModule(mlir::ModuleOp moduleOp) {
...
-  IREE_CHECK_OK(
-      iree_hal_driver_registry_allocate(iree_allocator_system(), &registry));
-  IREE_CHECK_OK(iree_hal_local_task_driver_module_register(registry));
-  IREE_CHECK_OK(iree_vm_instance_create(IREE_VM_TYPE_CAPACITY_DEFAULT,
-                                        iree_allocator_system(), &instance));
-  IREE_CHECK_OK(iree_hal_module_register_all_types(instance.get()));
+  if (iree_status_is_ok(initStatus))
+    initStatus =
+        iree_hal_driver_registry_allocate(iree_allocator_system(), &registry);
+  if (iree_status_is_ok(initStatus))
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/Runtime.cpp]: use {} when dealing with the runtime code, as some of these may be macros

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp [hunk 5]
```diff
@@ -472,17 +498,20 @@ InMemoryCompiledBinary::translateFromModule(mlir::ModuleOp moduleOp) {
...
+    initStatus = iree_hal_local_task_driver_module_register(registry);
+  if (iree_status_is_ok(initStatus))
+    initStatus = iree_vm_instance_create(IREE_VM_TYPE_CAPACITY_DEFAULT,
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/Runtime.cpp:524]: especially {} with multi-line

File: compiler/src/iree/compiler/ConstEval/test/jit_globals_vmvx.mlir [hunk 6]
```diff
@@ -0,0 +1,265 @@
+// RUN: iree-opt --split-input-file --iree-consteval-jit-globals --iree-consteval-jit-target-device=vmvx --iree-consteval-jit-debug --verify-diagnostics %s | FileCheck %s
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/test/jit_globals_vmvx.mlir]: why vmvx? you can have two RUN statements in the same file, one per backend
if we do have any different files it should just be for things that only work in one place or the other and all things expected to work everywhere (like all these) should be common

**Review by benvanik — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/ConstEval/test/jit_globals_vmvx.mlir (hunk 6 above)
Comment (ScottTodd) [reply compiler/src/iree/compiler/ConstEval/test/jit_globals_vmvx.mlir]: > I tried to keep both tests in the same file, but --verify-diagnostics isn't compatible with that style of lit testing AFAICT

I'll look more closely and iterate a bit. `--verify-diagnostics` checks the `// expected-warning` lines and those differ between vmvx and llvm-cpu. Wasn't sure if there was a way to mix those in a single file.

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp (hunk 3 above)
Comment (ScottTodd) [reply compiler/src/iree/compiler/ConstEval/Runtime.cpp]: https://github.com/openxla/iree/pull/16812 is still pending

File: compiler/src/iree/compiler/ConstEval/test/jit_globals.mlir [hunk 7]
```diff
@@ -142,14 +141,13 @@ module @eval_i1_tensor {
 }
 
 // -----
-// CHECK-LABEL: @eval_i4_tensor
+// expected-error @+2 {{OUT_OF_RANGE; attempted to access an address outside of the valid buffer range}}
 module @eval_i4_tensor {
   util.global private @hoisted : tensor<5x6xi4>
   util.func public @main() -> tensor<5x6xi4> {
     %hoisted = util.global.load @hoisted : tensor<5x6xi4>
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/ConstEval/test/jit_globals.mlir]: https://github.com/openxla/iree/actions/runs/8727055167/job/23943706482?pr=17075#step:4:12961 lol ASan doesn't like this test case either

```
==62069==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x511000191297 at pc 0x55622c1d52a4 bp 0x7ffe370e7c90 sp 0x7ffe370e7450
  WRITE of size 1 at 0x511000191297 thread T0
      #0 0x55622c1d52a3 in __asan_memcpy (/work/build-asan/tools/iree-opt+0xc42a3) (BuildId: 5c7c7a74b0ce49f09aceaca7b0b926e0f25945ff)
      #1 0x7f9fe253e78d in mlir::iree_compiler::IREE::Util::raw_inplace_ostream::write_impl(char const*, unsigned long) /work/compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.cpp:58:5
      #2 0x7f9fd9a75340 in llvm::raw_ostream::write(char const*, unsigned long) /work/third_party/llvm-project/llvm/lib/Support/raw_ostream.cpp:252:9
      #3 0x7f9fe25494e0 in mlir::iree_compiler::IREE::Util::serializeSplatValue(mlir::Location, mlir::Attribute, long, llvm::endianness, llvm::raw_ostream&) /work/compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.cpp:209:8
      #4 0x7f9fe25494e0 in mlir::iree_compiler::IREE::Util::SerializableDenseElementsAttrModel::serializeToStream(mlir::Attribute, mlir::Location, llvm::endianness, llvm::raw_ostream&) const /work/compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.cpp:795:14
      #5 0x7f9fe2548883 in mlir::iree_compiler::IREE::Util::SerializableDenseElementsAttrModel::serializeToBuffer(mlir::Attribute, mlir::Location, llvm::endianness, llvm::ArrayRef<char>) const /work/compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.cpp:782:12
      #6 0x7f9fe2548883 in mlir::iree_compiler::IREE::Util::detail::SerializableAttrInterfaceInterfaceTraits::FallbackModel<mlir::iree_compiler::IREE::Util::SerializableDenseElementsAttrModel>::serializeToBuffer(mlir::iree_compiler::IREE::Util::detail::SerializableAttrInterfaceInterfaceTraits::Concept const*, mlir::Attribute, mlir::Location, llvm::endianness, llvm::ArrayRef<char>) /work/build-asan/compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrInterfaces.h.inc:279:51
      #7 0x7f9fdca0055d in mlir::iree_compiler::ConstEval::FunctionCall::importSerializableAttr(mlir::Location, mlir::iree_compiler::IREE::Util::SerializableAttrInterface) /work/compiler/src/iree/compiler/ConstEval/Runtime.cpp:189:47
      #8 0x7f9fdca011b0 in mlir::iree_compiler::ConstEval::FunctionCall::addBufferViewArgumentAttr(mlir::Location, mlir::ShapedType, mlir::iree_compiler::IREE::Util::SerializableAttrInterface) /work/compiler/src/iree/compiler/ConstEval/Runtime.cpp:227:17
      #9 0x7f9fdca016b0 in mlir::iree_compiler::ConstEval::FunctionCall::addArgument(mlir::Location, mlir::Attribute) /work/compiler/src/iree/compiler/ConstEval/Runtime.cpp
```

Comment (ScottTodd) [reply compiler/src/iree/compiler/ConstEval/test/jit_globals.mlir]: Just reverted the changes from https://github.com/openxla/iree/pull/16321 - looks broken. Should I also revert the test cases that PR added while I'm at it?

File: compiler/src/iree/compiler/ConstEval/test/jit_globals_vmvx.mlir (hunk 6 above)
Comment (ScottTodd) [reply compiler/src/iree/compiler/ConstEval/test/jit_globals_vmvx.mlir]: Split the files a bit differently. Does this make sense? (Coverage isn't complete... would like a `vmvx-expected-warning` with `--verify-diagnostics` somehow...)

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp [hunk 8]
```diff
@@ -22,17 +23,8 @@ namespace {
...
-  if (!iree_status_format(status, message.size(), &message[0],
-                          &buffer_length)) {
-    message.resize(buffer_length + 1);
-    iree_status_format(status, message.size(), &message[0], &buffer_length);
-  }
-  message.resize(buffer_length);
-  iree_status_ignore(status);
-  return emitError(loc) << "runtime error in consteval: " << message;
+  return emitError(loc) << "runtime error in consteval: "
+                        << iree::Status::ToString(status);
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/ConstEval/Runtime.cpp]: FYI there was a memory issue in here that was crashing. The `iree::Status::ToString` helper is fine though.

**Comment by ScottTodd** [comment]
Ping?

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp (hunk 8 above)
Comment (benvanik) [reply compiler/src/iree/compiler/ConstEval/Runtime.cpp]: iree_status_ignore frees the status and is required not to leak memory - you need to have it - you just need to convert it to a string _before_ you free it

**Review by benvanik — CHANGES_REQUESTED** [review]
it'd be worth running asan given all the leaks :)

**Comment by ScottTodd** [comment]
> it'd be worth running asan given all the leaks :)

The tests are currently ASan-clean (they were not before I switched to use the helper function). I suppose I could run some of the new error handling paths through ASan manually too.

**Comment by ScottTodd** [comment]
Changed error handling, which fixed a few ASan issues reported when I forced some failures in the runtime functions used. PTAL?

File: compiler/src/iree/compiler/ConstEval/Runtime.h [hunk 9]
```diff
@@ -93,6 +96,7 @@ class Runtime {
 
   iree_hal_driver_registry_t *registry = nullptr;
   iree::vm::ref<iree_vm_instance_t> instance;
+  iree_status_t initStatus;
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/Runtime.h]: always default initialize things (here to iree_status_ok)

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp [hunk 10]
```diff
@@ -174,28 +178,39 @@ FunctionCall::importSerializableAttr(
...
+      status =
+          iree_make_status(IREE_STATUS_INTERNAL, "serializeToBuffer failed");
+    }
+  }
 
-  if (failed(handleRuntimeError(loc, iree_hal_buffer_unmap_range(&mapping))) ||
-      failed(copyResult)) {
-    return failure();
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_buffer_unmap_range(&mapping);
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/Runtime.cpp]: you want to unmap regardless of whether there was a failure - the only time you don't want to unmap is when the map failed - if you move this into the if above that does the serializeToBuffer that'll handle it

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp [hunk 11]
```diff
@@ -427,37 +446,58 @@ TypedAttr CompiledBinary::convertVariantToAttribute(Location loc,
...
-  IREE_CHECK_OK(iree_hal_driver_create_default_device(
-      driver, iree_allocator_system(), &device));
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_driver_registry_try_create(
+        runtime.registry, iree_make_cstring_view("local-task"),
+        iree_allocator_system(), &driver);
+  }
+
+  if (iree_status_is_ok(status)) {
+    iree_hal_driver_create_default_device(driver, iree_allocator_system(),
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/Runtime.cpp]: status=

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp [hunk 12]
```diff
@@ -427,37 +446,58 @@ TypedAttr CompiledBinary::convertVariantToAttribute(Location loc,
...
+  }
   iree_hal_driver_release(driver);
 
   // Create hal module.
-  iree_hal_device_t *device_ptr = device.get();
-  IREE_CHECK_OK(iree_hal_module_create(
-      runtime.instance.get(), /*device_count=*/1, &device_ptr,
-      IREE_HAL_MODULE_FLAG_NONE, iree_allocator_system(), &hal_module));
+  if (iree_status_is_ok(status)) {
+    iree_hal_device_t *device_ptr = device.get();
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/Runtime.cpp]: would be clearer if you did what's done below for modules (std::array and then passing array data()/size())

File: compiler/src/iree/compiler/ConstEval/Runtime.cpp [hunk 13]
```diff
@@ -472,20 +512,28 @@ InMemoryCompiledBinary::translateFromModule(mlir::ModuleOp moduleOp) {
...
 }
 
 Runtime::Runtime() {
-  IREE_CHECK_OK(
-      iree_hal_driver_registry_allocate(iree_allocator_system(), &registry));
-  IREE_CHECK_OK(iree_hal_local_task_driver_module_register(registry));
-  IREE_CHECK_OK(iree_vm_instance_create(IREE_VM_TYPE_CAPACITY_DEFAULT,
-                                        iree_allocator_system(), &instance));
-  IREE_CHECK_OK(iree_hal_module_register_all_types(instance.get()));
+  if (iree_status_is_ok(initStatus)) {
```
Comment (benvanik) [inline compiler/src/iree/compiler/ConstEval/Runtime.cpp:516]: because you don't default initialize initStatus this is wrong - you should initialize it

**Review by benvanik — CHANGES_REQUESTED** [review]

**Comment by ScottTodd** [comment]
Thanks. Addressed those latest comments.

**Review by benvanik — APPROVED** [review]

File: compiler/src/iree/compiler/ConstEval/JitGlobals.cpp [hunk 14]
```diff
@@ -603,12 +603,6 @@ struct JitGlobalsPass : public JitGlobalsBase<JitGlobalsPass> {
     SupportedFeatures s;
     Builder b(context);
 
-    // Exclude vmvx backend since there is no i4 support there causing
-    // the `eval_i4_tensor` test in `jit_globals.mlir` to fail.
-    // TODO(#16321): Enable on other backends once this has been tested
-    // outside llvm-cpu.
-    if (requestedTargetDevice == "llvm-cpu" && hasRequestedTargetDevice)
-      s.addScalarType(b.getIntegerType(4));
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/ConstEval/JitGlobals.cpp:611]: Tagging related issue:
* https://github.com/iree-org/iree/issues/15681

**Pull request merged**